flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/ArrayAggFunctionITCase.java

```diff
@@ -0,0 +1,95 @@
+
+import org.apache.flink.table.functions.BuiltInFunctionDefinitions;
+import org.apache.flink.types.Row;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.stream.Stream;
+
+import static org.apache.flink.table.api.DataTypes.ARRAY;
+import static org.apache.flink.table.api.DataTypes.INT;
+import static org.apache.flink.table.api.DataTypes.ROW;
+import static org.apache.flink.table.api.DataTypes.STRING;
+import static org.apache.flink.table.api.Expressions.$;
+import static org.apache.flink.types.RowKind.DELETE;
+import static org.apache.flink.types.RowKind.INSERT;
+import static org.apache.flink.types.RowKind.UPDATE_AFTER;
+import static org.apache.flink.types.RowKind.UPDATE_BEFORE;
+
+/** Tests for built-in ARRAY_AGG aggregation functions. */
+class ArrayAggFunctionITCase extends BuiltInAggregateFunctionTestBase {
+
+    @Override
+    Stream<TestSpec> getTestCaseSpecs() {
+        return Stream.of(
+                TestSpec.forFunction(BuiltInFunctionDefinitions.ARRAY_AGG)
+                        .withDescription("ARRAY changelog stream aggregation")
+                        .withSource(
+                                ROW(STRING(), INT()),
+                                Arrays.asList(
+                                        Row.ofKind(INSERT, "A", 1),
+                                        Row.ofKind(INSERT, "A", 2),
+                                        Row.ofKind(INSERT, "B", 2),
+                                        Row.ofKind(INSERT, "B", 2),
+                                        Row.ofKind(INSERT, "B", 3),
+                                        Row.ofKind(INSERT, "C", 3),
+                                        Row.ofKind(INSERT, "C", null),
+                                        Row.ofKind(INSERT, "D", null),
+                                        Row.ofKind(INSERT, "E", 4),
+                                        Row.ofKind(INSERT, "E", 5),
+                                        Row.ofKind(DELETE, "E", 5),
+                                        Row.ofKind(UPDATE_BEFORE, "E", 4),
+                                        Row.ofKind(UPDATE_AFTER, "E", 6)))
+                        .testResult(
+                                source ->
+                                        "SELECT f0, array_agg(f1) FROM " + source + " GROUP BY f0",
+                                TableApiAggSpec.groupBySelect(
+                                        Collections.singletonList($("f0")),
+                                        $("f0"),
+                                        $("f1").arrayAgg()),
+                                ROW(STRING(), ARRAY(INT())),
+                                ROW(STRING(), ARRAY(INT())),
+                                Arrays.asList(
+                                        Row.of("A", new Integer[] {1, 2}),
+                                        Row.of("B", new Integer[] {2, 2, 3}),
+                                        Row.of("C", new Integer[] {3}),
```

Comment:
I wonder why here is just 
```java
Row.of("C", new Integer[] {3}),
```
based on input I would expect
```java
Row.of("C", new Integer[] {3, null}),
```
or did I miss anything?

Reply:
Also similar query for postgres
```sql
with input(a, c) as (
select 'a', 1
union all 
select 'a', 2
union all 
select 'c', 3
union all 
select 'c', null
)
select a, array_agg(distinct c) from input group by a
```
gives
```
a|array_agg|
-+---------+
a|{1,2}    |
c|{3,NULL} |
```

Reply:
```sql
-- calcite
ARRAY_AGG([ ALL | DISTINCT ] value [ RESPECT NULLS | IGNORE NULLS ] [ ORDER BY orderItem [, orderItem ]* ] )
-- flink
ARRAY_AGG([ ALL | DISTINCT ] expression)
```

This function simplifies some aspects compared to calcite. 
1. Currently, the Flink parser does not support parsing RESPECT NULLS | IGNORE NULLS, but it is still possible to make changes to support it.
2. AggregateUtil#extractDistinctInformation 848~858 ignores the ignoreNulls fields. https://github.com/apache/flink/blob/6bdb4f752adb2b43dbadd8ad4fffcb4c00568dd3/flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/utils/AggregateUtil.scala#L848-L858
3. ListView does not support null values. https://github.com/apache/flink/blob/40cbf231e3dc15129b69136941adc15beebe8ef6/flink-table/flink-table-common/src/main/java/org/apache/flink/table/api/dataview/ListView.java#L45-L47

For that reasons I made some simplification. If we need it, I can make some attempts.

Reply:
It's ok to support it only partially however the main concern here: I would expect more consistent result
since there is also input `Row.ofKind(INSERT, "D", null),` and expected value `Row.of("D", null),`
So I would expect for ths either containing nulls or both not containing null however not mixed

Reply:
Thanks @snuyanzin, I'm trying to support `RESPECT NULLS | IGNORE NULLS` syntax.

Reply:
@snuyanzin 
`ListView<T>` cannot hold null values, and since it is a generic `List,` it cannot be represented by a specific object for null values. Is there any good solution for this?

https://github.com/apache/flink/blob/d92ab390ff0d91459daccfda7d6cb0acf6ca92eb/flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/aggregate/JsonArrayAggFunction.java#L65

Reply:
I think it's a very good point. It would be nice to be compatible with SQL and other vendors. 

One idea is we could keep the null indices and set those while retrieving the result. Happy to hear better solutions though.

Reply:
yes, I was also thinking about storing null indexes

Reply:
Hi @dawidwys, @snuyanzin,

I refers to `LagAggFunction` use `LinkedList` to hold null values and it works. Now `RESPECT NULLS` and `IGNORE NULLS` are supported. 
https://github.com/apache/flink/blob/1ffb481111f658b699702357921a48e914d13caf/flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/aggregate/LagAggFunction.java#L147-L150

Reply:
The downside of the solution is that the array must fit into memory at all times. The difference with the `LagAggFunction` is that `LAG` keeps at most `n` elements where `n` is controlled by the user.

Still I am reasonably good with the `LinkedList` approach because it anyhow needs to fit into memory when we emit it at the end as a single record. Writing this down for awareness.